File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays.DotNet/director/Chunks/RaysScriptChunk.cs
using ProjectorRays.Common;
using ProjectorRays.Director;
using ProjectorRays.LingoDec;

namespace ProjectorRays.director.Chunks;

public class RaysScriptChunk : RaysChunk
{
    public RaysCastMemberChunk? Member;
    public RaysScript Script;

    public RaysScriptChunk(RaysDirectorFile? dir) : base(dir, ChunkType.ScriptChunk)
    {
        Script = new RaysScript(dir?.Version ?? 0);
    }

    public override void Read(ReadStream stream)
    {
        Script.Read(stream);
    }

    public override void WriteJSON(RaysJSONWriter json)
    {
        json.StartObject();
        json.WriteField("scriptNumber", Script.ScriptNumber);
        json.WriteField("handlersCount", Script.HandlersCount);
        json.WriteKey("handlers");
        json.StartArray();
        foreach (var h in Script.Handlers)
            WriteHandlerJSON(h, json);
        json.EndArray();
        json.EndObject();
    }

    private static void WriteHandlerJSON(RaysHandler h, RaysJSONWriter json)
    {
        json.StartObject();
        json.WriteField("nameID", h.NameID);
        json.WriteField("argumentCount", h.ArgumentCount);
        json.WriteField("localsCount", h.LocalsCount);
        json.WriteField("globalsCount", h.GlobalsCount);
        json.EndObject();
    }
}

File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays_Python/director/chunks/rays_script_chunk.py
from .rays_chunk import RaysChunk, ChunkType
from ...common.stream import ReadStream

class RaysScriptChunk(RaysChunk):
    def __init__(self, dir=None):
        super().__init__(dir, ChunkType.ScriptChunk)
        self.script_data = b""
        self.member = None

    def read(self, stream: ReadStream):
        self.script_data = stream.read_bytes(len(stream.data) - stream.pos)

    def write_json(self, writer):
        writer.start_object()
        writer.write_key('length'); writer.write_val(len(self.script_data))
        writer.end_object()
